Unit tests + coverage reports, type checking, pre-commit, and requested fixes

2bndy5 opened the pull request:
## Issues addressed
- addresses #74 
- solves #75
- fixes #76
- resolves #77 
- partially-solves #73
   - Using the `database` option in the docker env is currently discouraged (until we can switch to a composite action).
      - The docker container (via clang-tidy) seems to be following a symlink `github/workspace` that leads to a inaccessible absolute path. This is preventing any compilation database from being properly used in the docker env.
   - The path provided to the `database` option is resolved into an absolute path (if it isn't already absolute). Thanks to @BurningEnlightenment for the suggestion!
   - I changed the src to output any error from clang-tidy despite the executable's return code. This sheds more light on if clang-tidy was able to use the specified path to a compilation database.

## And more
- skip executing clang-format if the `style` option is set to a blank string (`''`)
- allow any OS to accept a path to the installed tools via the `version` option. Previously, this was intended only for Windows OS, but it was needed to allow custom builds of the tools to be executed from a user-defined install path.
- switched to rely more on `pathlib` instead of `os` for file system operations. 
   - with the switch to using `pathlib.Path.rglob()` (instead of using `os.walk()`), I was able to remove a slightly misleading debug statement as well.
- moved install config to pyproject.toml, and updated the dockerfile accordingly. The setup.py file isn't actually used any more, but remains for older versions of pip.
- added numerous unit tests and integrated codecov.io for coverage reporting.
- simplified the stored list of files so that the same structure can be used for any _supported_ event
- improve translation from file's byte-offset to file's line & columns.
- ensure the demo C++ sources use LF despite the OS that's checking out the repo. This was needed to get consistent results from the unit tests about the previous point.
- added a CI workflow to publish to pypi or test-pypi. I also added the necessary pypi/test-pypi API tokens to this repo's secrets.
- added type checking from mypy tool
- added pre-commit config

2bndy5 commented:
![image](https://user-images.githubusercontent.com/14963867/183566965-148079cd-4f12-4c01-b9a7-c34c9c37c7ee.png)
And so begins the support testing for a PR event... This is why I can't have nice things.

2bndy5 commented:
Time to start testing PR events in the test-cpp-linter repo...

1. I'm mainly concerned with thread comments because direct usage of the REST API is not covered by the unit tests.
2. I also want to add this repo as a submodule to the test repo (for good measure).

shenxianpeng commented:
I still learn(review) the changes and then do some e2e tests, just want to make sure this PR will not cause regression because there are users using the master branch directly, like [google/flatbuffers](https://github.com/google/flatbuffers).

2bndy5 commented:
I added this repo as a submodule to the test repo and changed the CI workflows a bit (all changes are on the [test-backlogged-updates branch](https://github.com/cpp-linter/test-cpp-linter-action/tree/test-backlogged-updates)). But I haven't gotten around to testing it thoroughly yet.

shenxianpeng commented on cpp_linter/run.py:
```diff
@@ -115,17 +121,18 @@
     type=lambda input: input.lower() == "true",
-    help="Set this option to 'false' to analyse any source files in the repo. "
+    help="Set this option to 'false' to analyze any source files in the repo. "
     "Defaults to %(default)s.",
```
`files-changed-only` argument default value here is false but on README.md, it is noted 'ture'

2bndy5 replied:
This is so I can run the action locally without a docker. Using this feature currently requires a REST API token, and I haven't regenerated mine in a while. I suppose we could use git directly, but I'm not sure how to identify a commit that is in an open PR. So, if we switch to `git show`, we could only support push events on local machines.

2bndy5 replied:
Remember that all defaults from actions.yml are passed to the `cpp-linter` executable script when on a github runner. Unless users are running the action as a pkg in a github runner, then this may cause undesired results. The google/flatbuffers project was using the action.yml defaults.

shenxianpeng replied:
OK, I see

shenxianpeng replied:
> This is so I can run the action locally without a docker. 

This is very powerful. If possible I recommend users to use it as an action than a package, more users who used the used by section can display more count, then will attract more users to use it, the only drawback is it takes a little time than run as a package.

2bndy5 replied:
I'm trying to pivot away from the docker since it seems to have trouble reading the database and third party dependencies or using cross-compilers. See [github docs about creating a composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action). This will allow us to use clang-tools-pip + cpp-linter-action without a docker... But we're not quite there yet.

2bndy5 replied:
1. I'd like to start releasing this pkg to pypi first, so we can install the distributed wheel (faster than `pip install .`).
2. I'd also like to start hosting our own releases of clang-tools-static-binaries to ensure reliability.

We're going to have to make the transition to composite action in phases.

shenxianpeng replied:
I thinking about whether to create a new repository called `cpp-linter` which is the core repo of Org `cpp-linter`. Then move out the cpp_linter source from `cpp-linter-action` repository, only keep the necessary action part code, it will the action repo very simple.

The core repo supports GitHub API, but we are open to supporting GitLab, and Bitbucket if need to have it in the future,

shenxianpeng replied:
I agree with the two points you mentioned above. 
Again, it should be more good to release pkg to PyPI from the new repo `cpp-linter`

2bndy5 replied:
Yes abstracting the python source into a separate repo seems appropriate.

shenxianpeng commented on .github/workflows/run-dev-tests.yml:
```diff
@@ -0,0 +1,108 @@
+      matrix:
+        py: ['3.7', '3.8', '3.9', '3.10']
+        os: ['windows-latest', ubuntu-latest]
+        version: ['13', '12', '11', '10']
```
Do we need to include old versions like 9, 8, 7, etc for testing?

2bndy5 replied:
```suggestion
        version: ['13', '12', '11', '10', '9', '8', '7']
```

I only didn't add those because we never supported those versions with the docker (& it would mean longer workflow runs with them added).

2bndy5 replied:
the versions are now added

2bndy5 replied:
It looks like the YML output from clang-tidy had breaking changes in v9. Meaning our current code needs to support the older versions of clang-tidy's YML output.

This isn't actually used in deployment, but I set up unit tests to check if we can parse the YML output. This will be beneficial if we ever support posting suggestions from clang-tidy in the event's diff.

2bndy5 replied:
Versions 7 & 8 are now passing the unit tests. Historically, this is a new feature because we have not supported v7 & v8 in the docker.

I'd update the README about this, but the README is rather specific to using the docker. This new unit test workflow doesn't use the docker though.

shenxianpeng commented on .github/workflows/run-dev-tests.yml:
```diff
@@ -0,0 +1,114 @@
+    paths:
+      - "**.py"
+      - .pylintrc
+      - mypy.ini
```
mypy.ini seems to have been merged into `pyproject.toml`

2bndy5 replied:
oh yea. I'll change this to the pyproject.toml and pre-commit.yml

I consolidated linter configs into the toml to avoid repo clutter. I want is people to see the README when they browse the root folder on the website. If there are too many folder/files in the root folder then they have to scroll down to see the README... It looks like I can hide the [coverage config in the toml](https://coverage.readthedocs.io/en/latest/config.html) also.

2bndy5 replied:
I updated these triggers.

I moved the coverage config to the toml. I also added a basic pytest config to the toml, so it doesn't need any args passed to the `pytest` command.

2bndy5 replied:
With all the new hooks added to pre-commit, it might be better to just run this workflow regardless of the paths. Currently, the yaml and requirements.txt files aren't getting checked in the CI unless there was also changes made to
https://github.com/cpp-linter/cpp-linter-action/blob/23622e1efccbff34aa4ab0b299bef38deb127a5f/.github/workflows/run-dev-tests.yml#L5-L9

2bndy5 replied:
Nevermind, I just moved the pre-commit steps into its own CI workflow.

shenxianpeng commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,26 @@
+repos:
+  - repo: https://github.com/python/black
```
How about adding the following common hooks to `.pre-commit-config.yaml`

```yml
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.3.0
    hooks:
      - id: trailing-whitespace
      - id: end-of-file-fixer
      - id: check-docstring-first
      - id: check-added-large-files
      - id: check-yaml
      - id: check-toml
      - id: requirements-txt-fixer
```

2bndy5 replied:
added

shenxianpeng reviewed (approved):
I'm not very familiar with some of the specific details, so I can't give any advice, but I feel great :) If both the Unit test and the test action pass, it looks good to me to merge.

2bndy5 commented:
The file annotations seem to be in order also (see [here](https://github.com/cpp-linter/cpp-linter-action/pull/81/commits/31c332fab4f67d03a86d46d12163764824e4b838#diff-513714f4594a117388fca28e69bd00b1b991e0ae9637f2df95a4b3c4f35c6098)). I'm much more confident now that we have 80% coverage and expected behavior in the unit tests.

shenxianpeng commented:
Could you also please update README.md to introduce `v2` in this PR or #83

2bndy5 commented:
I think the `v2` should go in #83 All the changes here were made with backward compatibility in mind. I do think there's some changes that should be made regarding the input options, so I'll re-read that file for accuracy.

shenxianpeng commented:
Sure, please go ahead with #83

2bndy5 commented:
Is the .ci-ignore file still needed? https://github.com/cpp-linter/cpp-linter-action/blob/a1787309cce72d953c7bae2b3d25f95140c99268/.ci-ignore#L1-L2

I recall this was for a external linter CI you had configured for the repo, but not sure if changing to an org account affected that.

shenxianpeng commented:
I forgot it was configured by me. I even don't know what it used for, I think we don't need it

2bndy5 commented:
I think I'll leave this alone until I'm sure the pkg is available on pypi

> https://github.com/cpp-linter/cpp-linter-action/blob/a1787309cce72d953c7bae2b3d25f95140c99268/README.md#L185

It would be nice to have all versions of the python source code up on pypi, but that's a lot of work for code that is known to be flawed.

2bndy5 commented:
Please look over the readme changes. I also removed that .ci-ignore file.

2bndy5 commented:
All good to go on my end. Here's hoping uploads to pypi don't have the same problems we experienced with cpp-linter/clang-tools-pip v0.3.0 release. 🤞🏼

shenxianpeng commented:
Why not add PyPI token to cpp-linter repo, do you want to upload the package from this repo first and then migrate it to cpp-linter?

I used the same approach(in clang-tools-pip) and upload cpp-linter/cpp-linter-hooks to PyPI, everything goes well.

2bndy5 commented:
> do you want to upload the package from this repo first and then migrate it to cpp-linter?

Yes. That would eliminate some guesswork from migration. PyPI doesn't care about the origin as long as the right credentials are used.

2bndy5 commented:
According to [the codecov action](https://github.com/codecov/codecov-action#arguments), the secret token isn't needed for public repos.

shenxianpeng commented:
OK, please go ahead to remove the token.

2bndy5 commented:
Done.

If you're using the upload script from codecov's instruction on the cpp-linter-hook repo, then switching to the codecov action will eliminate the need for that token in the org secrets.

I'm honestly not sure if the token is needed after uploading the first report.

shenxianpeng commented:
I remember I used the upload script and token because it was shown in the example on codecov portal. It's Okay to remove the token and replaced with action, it should be easy to fix if any errors report.

2bndy5 commented:
done.

shenxianpeng commented:
> Done (with no problems).

So ready to merge?

2bndy5 commented:
yep
